### ISSUE TITLE
Crash when adding column

### ISSUE THREAD

**Issue by JoopClaireIT** [alexbrainman/odbc#62]
![screenshot 2015-10-29 at 16 55 34](https://cloud.githubusercontent.com/assets/13216791/10823884/0b89629a-7e5e-11e5-95c3-3ef1eeb9d4f1.png)

Sorry, didn't manage to copy/paste the error in this case. Will try to debug and solve myself, then issue PR, but just so you know about it :smiley:.

**Comment by alexbrainman** [comment]
If you can provide me with a small program I can run myself, I will investigate. Otherwise it is hard to guess what the problem is.

Alex

**Comment by JoopClaireIT** [comment]
I know, will have to narrow it down myself first...

Op vr 30 okt. 2015 00:15 schreef Alex Brainman <notifications@github.com>:

> If you can provide me with a small program I can run myself, I will
> investigate. Otherwise it is hard to guess what the problem is.
>
> Alex
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/alexbrainman/odbc/issues/62#issuecomment-152353895>.
>
-- 
Met vriendelijke groet,

Joop Kiefte

Claire Automotive Support - Galvanistraat 1 - 6716AE - EDE
Tel 0850020211 - www.claireit.nl

**Comment by JoopClaireIT** [comment]
It only happens with one specific column, it's not number of columns, I bet it's something with the column type. This is with an Intersystems Caché database (uses SQL abstracting over an OO system internally, and technically only using strings internally). The contents of the specific column is a string describing car types. The specific lines point to the `Next()` method, and there in the columns.go file to the `return c.BaseColumn.Value(c.Buffer[:c.Len])` line. So maybe in some corner cases c.Len is not accurate?

**Comment by JoopClaireIT** [comment]
Just logged it with a printf of len(c.Buffer) and c.Len just before the return, and in the loop it goes well most of the time, then produces len(c.Buffer) as 51 but c.Len as 60. Maybe there needs to be an additional bounds check here?

**Comment by alexbrainman** [comment]
I agree it must be to do with the way ODBC driver reports your column types. Is it possible for me reproduce this bug here? What software do I need? What are the steps? This package has been mainly used with MS SQL Server, and I wouldn't be surprised if there is some missing logice for other databases.

Alex

**Comment by JoopClaireIT** [comment]
The database is Intersystems Caché, it's a commercial one so not sure if you would be able to gain access to one. The ODBC driver is a free download though, and there seems to be a free full functioning one person trial without time limit or other limitations that would make debugging impossible... The exact bug happened with me only in a rare case but on that specific case always, not intermittently, so I think you should be able to trace the origin then...

**Comment by ariel17ml** [comment]
Hello @alexbrainman. I've faced same problem using this driver with Apache Hive (query with single string column), applied changes provided by @JoopClaireIT and now it works like expected. Is there any planned ETA for merging it into production?

Thank you and best regards.
Ariel Ríos.

**Comment by alexbrainman** [comment]
> Is there any planned ETA for merging it into production?

I would like to understand what is the problem that this PR fixes first.

Can you provide a small program that demonstrates the problem? Run your program, show its output here, and explain to me why you think that your program output is wrong.

Can I run your program here, on my own computer? What do I need to run it?

Thank you.

Alex

**Comment by haynesherway** [comment]
I am having the same issue, specifically it seems with null dates coming from an IBM iSeries database. They don't report SQL_NULL_DATA as true, so it doesn't get caught with the IsNull(). Then it panics when this slice is out of range. I spent way too much of my life investigating this.
